### DISCUSSION TITLE
Please make it possible to create custom tooltips with C# source code on Editor.

### DISCUSSION THREAD

**Discussion by UdonSoft** [godotengine/godot-proposals#5675] [Editor]
### Describe the project you are working on

Because we are creating a game that depends on physics,
Creating a physics-based character controller

Therefore, the amount of code will inevitably increase and the amount of variables exposed on the Editor will also increase.

### Describe the problem or limitation you are having in your project

If you can't create a Tooltip that can be displayed on the Editor
I need to open the IDE to see what the variables mean
very troublesome

GDScript allows you to create custom Tooltips
I can't believe it's not possible in C#.

![CSharp_Tooltip](https://user-images.githubusercontent.com/73213136/198808376-8e0bce55-5934-4056-bd47-638b1e6ff0e0.png)
![GDScript_Tooltip](https://user-images.githubusercontent.com/73213136/198808388-e1815a28-3a31-4774-9ae2-a7315df5e327.png)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Create custom tooltips in C# code
Please make it possible.

This allows you to check the meaning of the variables on the Editor, so
Increased development iterations

In team development
For artists and designers adjusting variables
It has great benefits.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Using doc comments in C#
Display a custom tooltip on the Editor

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Currently there is no solution.

### Is there a reason why this should be core and not an add-on in the asset library?

as explained above
・Improve iteration
・Checking variables in team projects
There are many advantages

**Comment by geekley** [comment]
Is worth thinking whether we should just automatically use XMLDoc assuming all code documentation in classes and exported members is always desired as tooltip, or if we should require explicitly enabling that. I don't know if there are valid situations where one would want different doc in code and editor, but it could be the case ...?

For example, through special markup:
```cs
///<tooltip><summary>Tooltip blah blah.</summary></tooltip>
///<remarks>Not in tooltip.</remarks>
[Export] string myProp;
```
... or maybe even an attribute:
```cs
///<summary>Tooltip blah blah.</summary>
[Export, Tooltip] string myProp;
///<summary>Not in tooltip.</summary>
[Export, Tooltip("Tooltip blah blah.")] string otherProp;
```
I'm not advocating for either solution btw, just wanted to throw an idea here. I'm fine with any implementation, and if all XMLDoc in exported members is automatically tooltip, that's fine by me too. Might even be the best solution.

In any case, yes, please do bring this GDScript feature to C# as well! :smile:

**Reply by blueknightone** [reply]
I think it would be pretty safe to assume that if someone put an XMLDoc block on top of an exported member, they would expect that to come over into the Inspector without any extra steps or nesting blocks. I would say just have the `<summary>` block and ignore anything else.

It would also be extra motivation for people like me to be better about documenting their code. Seeing "No description." on a tooltip for something I have control over makes my eye twitch a little.

**Comment by cai-tan** [comment]
Why is this a discussion and not an issue? Isn't this a missing feature?

**Reply by Calinou** [reply]
Proposal issues are meant to be used when you have a clear, concrete idea of how the feature should work in practice (and ideally, how to implement it in the engine source code). Proposal issues are recommended for those who intend to implement the feature they're proposing themselves.

If you don't have a clear idea of the implementation details, discussions are more suited as they are designed to be open-ended :slightly_smiling_face:

**Reply by cai-tan** [reply]
I'm pretty sure the original issue here definitely outlined exactly the expected functionality of the feature: Use the XML documentation `<summary>` block that is already a feature of C# entirely analogous to the `##` blocks in GDScript. 

You say "a clear idea of the implementation details." How much of the underlying engine code do we need to be aware of to make a proposal issue on something so basic? Am I better off spending eight hours forking the full source code of the engine and submitting a pull request than even bothering with a proposal issue?

**Comment by jorammer** [comment]
I am new to Godot (Unity refugee). This post is a year old. I think it's pretty clear what people want (and it already works with GDScript). Could this be converted to an issue so this can be on some sort of road map? (I noticed that I can convert this to an issue myself, but -as I noted earlier- I am new to this community, so I don't feel it's my place to start raising issues)

**Comment by Calinou** [comment]
Closing in favor of https://github.com/godotengine/godot-proposals/issues/8269, as a technical implementation was provided in https://github.com/godotengine/godot-proposals/issues/8269#issuecomment-1782701674.
